File: Data Structures & Algorithms/subarray-sum-equals-k/submission-2.kt
class Solution {
    fun subarraySum(nums: IntArray, k: Int): Int {

        var map = HashMap<Int, Int>()
        map[0] = 1
        var output = 0
        var sum = 0
        for (i in 0 until nums.count()) {
            sum += nums[i]
            if (map[sum - k] != null) {
                output += map[sum - k]!!
            }
            map[sum] = (map[sum] ?: 0) + 1
        }
        return output
    }
}

File: Data Structures & Algorithms/subsets-ii/submission-2.swift
class Solution {
    func subsetsWithDup(_ nums: [Int]) -> [[Int]] {
        var sorted = nums.sorted()
        var output = [[Int]]()
        var temp = [Int]()

        func dfs(_ i: Int) {
            if i == sorted.count {
                output.append(temp)
                return
            }
            // select i
            temp.append(sorted[i])
            dfs(i + 1)

            temp.removeLast()
            var j = i + 1
            while j < sorted.count && sorted[j] == sorted[j - 1] {
                j += 1
            }
            dfs(j)
        }

        dfs(0)

        return output
    }
}
